source/MaterialXGraphEditor/Graph.cpp

```diff
@@ -62,32 +62,44 @@ Graph::Graph(const std::string& materialFilename,
     loadStandardLibraries();
     setPinColor();
 
-    // Generate node UI from nodedefs.
-    std::vector<mx::NodeDefPtr> nodeDefs = _stdLib->getNodeDefs();
-    for (size_t i = 0; i < nodeDefs.size(); i++)
+    _graphDoc = loadDocument(materialFilename);
+    _graphDoc->importLibrary(_stdLib);
+
+    _initial = true;
+    createNodeUIList(_stdLib);
+
+    if (_graphDoc)
     {
-        // nodeDef is the key for the map
-        std::string group = nodeDefs[i]->getNodeGroup();
-        if (group == "")
-        {
-            group = "extra";
-        }
-        std::unordered_map<std::string, std::vector<mx::NodeDefPtr>>::iterator it = _nodesToAdd.find(group);
-        if (it == _nodesToAdd.end())
-        {
-            std::vector<mx::NodeDefPtr> nodes;
-            _nodesToAdd[group] = nodes;
-        }
-        _nodesToAdd[group].push_back(nodeDefs[i]);
+        buildUiBaseGraph(_graphDoc);
+        _currGraphElem = _graphDoc;
+        _prevUiNode = nullptr;
     }
 
+    // Create a renderer using the initial startup document.
+    // Note that this document may have no nodes in it
+    // if the material file name does not exist.
     mx::FilePath captureFilename = "resources/Materials/Examples/example.png";
     std::string envRadianceFilename = "resources/Lights/san_giuseppe_bridge_split.hdr";
-    mx::Color3 screenColor(1.0f, 0.3f, 0.32f);
-
-    _renderer = std::make_shared<RenderView>(materialFilename, meshFilename, envRadianceFilename,
-                                             _searchPath, _libraryFolders, 256, 256);
+    _renderer = std::make_shared<RenderView>(_graphDoc, meshFilename, envRadianceFilename,
```

Comment:
Instead of duplicate loading of documents + libraries, the Graph is responsible for this and passes this on to the renderer. The renderer can add additional includes which are parsed to be part of the filtering on save as well as when building the node graph to display.